File: tests/dropout.cs
using System;
using System.IO;

using nn;

unsafe internal static class dropout {
    public static void test_dropout_forward(TextWriter Console, double p, bool train) {
        Console.WriteLine($"<test_dropout_forward p={p:f1} mode='{(train ? "train" : "eval")}'>");
        var g = new nn.rand.mt19937(137);
        var a = Tensor.NaN(17);
        Console.WriteLine(g.randint32());
        nn.rand.normal_(a.data, a.numel(), g, 0, 1);
        Console.WriteLine(Common.pretty_logits(a.data, a.numel()));
        Console.WriteLine(g.randint32());
        var mask = Tensor.NaN(7);
        F.dropout_forward_cpu(
            a.data,
            a.data,
            mask.data,
            a.numel(),
            p,
            train,
            g);
        Console.WriteLine(Common.pretty_logits(a.data, a.numel(), 0xFFFFFFFF));
        Console.WriteLine(g.randint32());
        Console.WriteLine("</test_dropout_forward>");
    }

    public static void test_dropout_backward(TextWriter Console, double p, bool use_dropout) {
        Console.WriteLine($"<test_dropout_backward p={p:f1} dropout='{(use_dropout ? "yes" : "no")}'>");
        var g = new nn.rand.mt19937(137);
        Console.WriteLine(g.randint32());
        Linear hidden = new nn.Linear(7, 7, bias: false);
        nn.init.kaiming_uniform_(
            hidden._Weight.data,
            hidden._Weight.numel(),
            g,
            hidden.I,
            (float)Math.Sqrt(5));
        Dropout dropout = use_dropout ? new Dropout(g, p) : null;
        Tanh tanh = new nn.Tanh();
        Console.WriteLine(g.randint32());
        Console.WriteLine("weight:");
        Console.WriteLine(Common.pretty_logits(hidden._Weight.data, hidden._Weight.numel(), 0xFFFFFFFF));
        var input = Tensor.zeros(7, requires_grad: true);
        nn.rand.normal_(input.data, input.numel(), g, 0, 1);
        Console.WriteLine(g.randint32());
        var output = hidden.forward(input);
        if (dropout != null) {
            output = dropout.forward(output);
        }
        output = tanh.forward(output);
        Console.WriteLine("output:");
        Console.WriteLine(Common.pretty_logits(output.data, output.numel(), 0xFFFFFFFF));
        Console.WriteLine(g.randint32());
        var target = Tensor.ones(7);
        var loss = F.mse_loss(
            output.data,
            output.grad,
            output.numel(),
            target.data);
        Console.WriteLine("loss:");
        Console.WriteLine($"[{loss:f4}]");
        output = tanh.backward(output);
        if (dropout != null) {
            output = dropout.backward(output);
        }
        hidden.backward(output);
        Console.WriteLine(g.randint32());
        Console.WriteLine("weight.grad:");
        Console.WriteLine(Common.pretty_logits(hidden._Weight.grad, hidden._Weight.numel(), 0xFFFFFFFF));
        Console.WriteLine(g.randint32());
        Console.WriteLine("</test_dropout_backward>");
    }
}

File: tests/dropout.py
import numpy as np;

import torch
import torch.nn as nn
import torch.nn.functional as F

from common import pretty_logits

def test_dropout_forward(p, train):
    print("<test_dropout_forward p=" + str(p) + " mode='" + ("train" if train else "eval") + "'>")
    torch.manual_seed(137)
    print(torch.randint(0, 0xFFFFFFFF, [1]).item())
    model = nn.Dropout(p)
    if train:
        model.train();
    else:
        model.eval();
    input = torch.randn(17)
    print(pretty_logits(input))
    print(torch.randint(0, 0xFFFFFFFF, [1]).item())
    output = model(input);
    print(pretty_logits(output, 0xFFFFFFFF))
    print(torch.randint(0, 0xFFFFFFFF, [1]).item())
    print("</test_dropout_forward>")

class Net(nn.Module):
    def __init__(self, input_size, num_classes, dropout, p):
        super(Net, self).__init__()
        self.hidden = nn.Linear(input_size, num_classes, bias=False)
        self.dropout = None
        if dropout:
            self.dropout = nn.Dropout(p)
        self.tanh = nn.Tanh()

    def forward(self, x):
        x = self.hidden(x)
        if self.dropout is not None:
            x = self.dropout(x)
        x = self.tanh(x)
        return x

def test_dropout_backward(use_dropout, p):
    print("<test_dropout_backward p=" + str(p) + " dropout='" + ("yes" if use_dropout else "no") + "'>")
    torch.manual_seed(137)
    print(torch.randint(0, 0xFFFFFFFF, [1]).item())
    model = Net(7, 7, use_dropout, p)
    model.train()
    print(torch.randint(0, 0xFFFFFFFF, [1]).item())
    print("weight:")
    print(pretty_logits(model.hidden.weight, 0xFFFFFFFF))
    input = torch.randn(7)
    print(torch.randint(0, 0xFFFFFFFF, [1]).item())
    output = model(input)
    print("output:")
    print(pretty_logits(output, 0xFFFFFFFF))
    print(torch.randint(0, 0xFFFFFFFF, [1]).item())
    target = torch.ones(7)
    loss = F.mse_loss(output, target)
    print("loss:")
    print(pretty_logits(loss))
    loss.backward()
    print(torch.randint(0, 0xFFFFFFFF, [1]).item())
    print("weight.grad:")
    print(pretty_logits(model.hidden.weight.grad, 0xFFFFFFFF))
    print(torch.randint(0, 0xFFFFFFFF, [1]).item())
    print("</test_dropout_backward>")

if __name__ == "__main__":
    import argparse

    from pathlib import Path

    path = Path(__file__).parent.absolute()

    parser = argparse.ArgumentParser()
    parser.add_argument("--p", type=float, default=0.5)
    parser.add_argument("--mode", type=str, default="train")
    parser.add_argument("--net", type=str, default="no")
    parser.add_argument("--dropout", type=str, default="yes")
    args = parser.parse_args()
    
    assert args.mode in {"train", "eval"}
    assert 0 <= args.p <= 1
    assert args.net in {"yes", "no"}
    assert args.dropout in {"yes", "no"}
    
    if args.net == "no":
        if args.mode == "train":
            test_dropout_forward(args.p, True);
        elif args.mode == "eval":
            test_dropout_forward(args.p, False);
    elif args.net == "yes":
        if args.dropout == "yes":
            test_dropout_backward(use_dropout = True, p = args.p);
        elif args.dropout == "no":
            test_dropout_backward(use_dropout = False, p = args.p);
